torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, ncclComm_t*> _communicators;
```

Comment:
Use `at::IntList` as a key (and add a custom hasher if it doesn't work out of the box)

Reply:
Also, you can use an `std::unique_ptr` with a deleter that calls `ncclComDestroy`. You don't need `THCPModule_nccl_destroy` at all if you do this

Reply:
you cant use at::IntList. IntList is of type ArrayRef, which is a loose wrapper around an array. ArrayRef does not own it's data and is not safe to store.

```
  /// This class does not own the underlying data, it is expected to be used in
  /// situations where the data resides in some other buffer, whose lifetime
  /// extends past that of the ArrayRef. For this reason, it is not in general
  /// safe to store an ArrayRef.
```